lib/cppapi/include/atscppapi/Async.h

```diff
@@ -209,13 +213,20 @@ template <typename AsyncProviderType> class AsyncReceiver : noncopyable
    */
   virtual void handleAsyncComplete(AsyncProviderType &provider) = 0;
   virtual ~AsyncReceiver() {}
+  void
+  revokePromise(AsyncDispatchController<AsyncReceiver<AsyncProviderType>, AsyncProviderType> *dispatch_controller_ptr)
+  {
+    receiver_promises_.erase(dispatch_controller_ptr);
+  }
 
 protected:
   AsyncReceiver() {}
   friend class Async;
 
 private:
-  mutable std::list<std::shared_ptr<AsyncReceiverPromise<AsyncReceiver<AsyncProviderType>, AsyncProviderType>>> receiver_promises_;
+  mutable std::unordered_map<AsyncDispatchController<AsyncReceiver<AsyncProviderType>, AsyncProviderType> *,
```

Comment:
Do you need `unordered_map`? Could you use `unordered_set` with a specialized hash that extracts the raw pointer from the `shared_ptr`?